src/storage.ts

```diff
@@ -1,146 +1,182 @@
+        const { movie_id } = await this.getMovieById(movieId);
+
+        if ( isFavorite ) {
+            await this.db.query(
+                `
+                INSERT INTO
+                    favorites (user_id, movie_id)
+                VALUES ($1, $2)
+                ON CONFLICT (user_id, movie_id)
+                DO UPDATE SET user_id = $1, movie_id = $2
+                `, [user_id, movie_id]
+            );
+        } else {
+            await this.db.query('DELETE FROM favorites WHERE user_id = $1 AND movie_id = $2', [user_id, movie_id]);
+        }
+    }
+
+    async addMovie ( data: any, email: string ) {
         const { imdbID: id, name } = data;
 
-        if ( !id && !name ) {
-            return false;
-        }
+        const { user_id } = await this.findUser(email);
+        const queryResult = await this.db.query(
+            "SELECT * FROM movies WHERE data ->> 'imdbID' = $1 OR data ->> 'name' = $2",
+            [id, name]
+        );
 
-        let index = id
-            ? this.__store.movies.findIndex(item => item.imdbID === id)
-            : this.__store.movies.findIndex(item => item.name === name);
+        let movieId;
 
-        index === -1
-            ? this.__store.movies.push(data)
+        if ( queryResult.rowCount === 0 ) {
+            movieId = await this.db.query('INSERT INTO movies (data, user_id) VALUES ($1, $2) RETURNING movie_id', [data, user_id]);
+        } else {
             // data can be updated, for example rating, votes, awards etc., so update in storage
-            : (this.__store.movies[index] = data);
+            movieId = await this.db.query('UPDATE movies SET data = $1 WHERE movie_id = $2 RETURNING movie_id', [data, queryResult.rows[0].movie_id]);
+        }
 
-        return data;
+        return movieId.rows[0].movie_id;
     }
 
-    read ( id: string | null, filters?: string ) {
-        const payload = id
-            ? cloneDeep(this.__store.movies.filter(item => item.imdbID === id))
-            : cloneDeep(this.__store.movies);
+    async getMovies ( id: string | null, filters?: string ) {
+        let queryResult;
 
-        return filters ? filterMovieAttributes(payload, filters) : payload;
-    }
+        if ( id ) {
+            queryResult = await this.db.query(
+                `
+                    SELECT COALESCE(movies.data || meta.data, movies.data)
+                    FROM movies
+                    LEFT JOIN meta ON movies.movie_id = meta.movie_id
+                    WHERE movies.data ->> 'imdbID' = $1 OR movies.data ->> 'name' = $1
+                `, [id]
+            );
 
-    update ( id: string, data: any ) {
-        const entry = this.__store.movies.find(item => item.imdbID === id);
+            return queryResult.rows[0].coalesce;
+        }
 
-        if ( entry ) {
-            const { comment, personalScore } = data;
+        queryResult = await this.db.query('SELECT movies.data FROM movies FULL JOIN meta ON movies.movie_id = meta.movie_id', []);
```

Comment:
why do you need full join if you only take data from movies table

Reply:
Absolutely, I need all movie records, but some of movies can had comments (so called metadata) from another table. So, left join is appropriate here. Thanks! Fixed.